src/editor fullscreen.js

```diff
@@ -0,0 +1,94 @@
+why 'enter' is the default fullscreen hotkey, to allow toggling fullscreen from a gamepad.
+
+
+HOW TO USE:
+1. Import this plugin into your game.
+2. Run the game.
+3. Press the 'enter' key.  Note that the game becomes fullscreen.
+4. Press the 'enter' key again.  Note that the game becomes non-fullscreen.
+5. Find the fullscreen button on the lower left of the ui (while the game is running).
+6. Press the fullscreen button.  Note that the game becomes fullscreen.
+NOTE: if this plugin's configuration is changed, the key to press may not be 'enter'.
+
+
+// Which keyboard key triggers fullscreen
+//!CONFIG fullscreen-hotkey (text) "Enter"
+*/
+
+//! CODE_EDITOR
+
+function setupFullscreenToggle() {
+	const KEY_FOR_FULLSCREEN = FIELD(CONFIG, 'fullscreen-keycode', 'text') || 'Enter';
+
+	// Fullscreen toggle logic
+	window.EDITOR.toggleFullscreen = function toggleFullscreen() {
+		const canvas = this.playtestIframe.contentWindow.document.getElementById('player-canvas');
+		if (!canvas) {
+			return;
+		}
+		if (!document.fullscreenElement && !document.mozFullScreenElement && !document.webkitFullscreenElement && !document.msFullscreenElement) {
+			if (canvas.requestFullscreen) {
+				canvas.requestFullscreen();
+			} else if (document.documentElement.msRequestFullscreen) {
+				canvas.msRequestFullscreen();
+			} else if (document.documentElement.mozRequestFullScreen) {
+				canvas.mozRequestFullScreen();
+			} else if (document.documentElement.webkitRequestFullscreen) {
+				canvas.webkitRequestFullscreen(Element.ALLOW_KEYBOARD_INPUT);
+			}
+		} else {
+			// eslint-disable-next-line no-lonely-if
+			if (document.exitFullscreen) {
+				document.exitFullscreen();
+			} else if (document.msExitFullscreen) {
+				document.msExitFullscreen();
+			} else if (document.mozCancelFullScreen) {
+				document.mozCancelFullScreen();
+			} else if (document.webkitExitFullscreen) {
+				document.webkitExitFullscreen();
+			}
+		}
+	};
+
+	// Playtest toolbar button
+	const playtestToolbar = document.getElementById('play-tab-view').children[1];
+	const fullscreenButton = document.createElement('button');
+	playtestToolbar.prepend(fullscreenButton);
+	fullscreenButton.id = 'fullscreenButton';
+	fullscreenButton.classList.add('icon-button');
+	fullscreenButton.title = 'toggle fullscreen';
+	fullscreenButton.innerHTML = `<svg xmlns="http://www.w3.org/2000/svg" width="512" height="512" viewbox="0 0 16 16" stroke="currentColor" stroke-linecap="round" fill="none" stroke-width="1.25"><path d="M5 2h-3v3"/><path d="M11 2h3v3"/><path d="M2 11v3h3"/><path d="M14 11v3h-3"/></svg>`;
+	fullscreenButton.addEventListener('click', window.EDITOR.toggleFullscreen.bind(window.EDITOR));
+
+	// Hotkey
+	document.addEventListener('keydown', e => {
+		if (e.key === KEY_FOR_FULLSCREEN) {
+			window.EDITOR.toggleFullscreen();
+		}
+	});
+}
+
+// Setup the plugin, if not already setup
+if (!document.getElementById('fullscreenButton')) {
+	setupFullscreenToggle();
+}
```

Comment:
noticed a minor bug here: because you skip the skip the setup if the plugin has already been setup, any changes to `fullscreen-hotkey` are not reflected until the page is refreshed (the event listener will have already been setup as "Enter" and won't be updated with the new config)

i'd recommend changing this to either delete + recreate the button, or remove + re-add the event listener, or read the key from the config instead of a const inside the event listener so it doesn't need to be reconfigured

Reply:
FYI: allowing dynamic hotkey updates without refreshing the page is going to require a modification to the editor-plugin system to provide event plugins with access to their event ids.  It's a simple change, but needs a PR to bipsi before I can make this work.  It might take a bit.

Reply:
maybe i'm missing something, but it looks like the hotkey is handled as a standard event listener and doesn't have any dependency on how bipsi handles things? and as mentioned above, you could fix this by retrieving the key inside the listener and not have to worry about removing/re-adding it

Reply:
The issue is fixed in the latest commit as soon as the bipsi PR is accepted.

> you could fix this by retrieving the key inside the listener

The problem is that the CONFIG event is created when the plugin is first loaded.  At that time, the plugin event's fields are copied into CONFIG.  Therefore, CONFIG doesn't reflect real-time changes to the plugin's fields.  This has been fine up to now, since the plugin is re-loaded for each game run, and you aren't going to change the plugin settings mid-playtest.  Now, with editor plugin, it's better to take the settings directly from the event, instead of the CONFIG event.

Reply:
i think you've overcomplicated the problem: when you do
```js
const foo = FIELD(CONFIG, 'bar');
```
this creates a variable with the value of "the field `bar` on the plugin's config. This value is a string primitive, so it will never change even if the plugin config is updated. However, `CONFIG` should be a reference to the plugin's config (not a copy of its values at the time of execution) and you should be able to use it to retrieve the up-to-date field value at any point

taking the settings from the event like you've done here is just adding extra indirection that should not be necessary

Reply:
The way CONFIG is handled makes sense when the game is exported: the values are hard-coded into the webpage and immediately available on execution.  Making it a reference to the event if created outside of the playtest has one issue that I can see: the undo/redo feature can change the current event object being used for the given event id.

One solution would be to make CONFIG a blank object with a prototype to the event it represents.  Then, each undo/redo can reassign the prototype.  I'll look into this as a viable option.

Reply:
... it might be a lot of work, but could still be worth it to allow the plugin developer to work with CONFIG the same way for all plugin modes.

Reply:
FYI, I implemented CONFIG as an empty object with the event as a prototype, like described above.  It actually wasn't that bad, and it works really well.  I'm waiting on Candle to approve the pull request.

Reply:
Ok.  Bipsi is now updated so that CONFIG is dynamic for editor plugins.  I've updated this plugin to reference that directly.